## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics
description= df.describe()
print(description)

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [5]:
# What three item categories had the most entries?
categories_highest = df['category'].value_counts()
sorted_categories = categories_highest.sort_values(ascending=False)
top_3_categories = sorted_categories.head(3)
print ("Top three item categories with most entries:")
print(top_3_categories)


Top three item categories with most entries:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [6]:
# For the category with the most entries, which subcategory had the most entries?
subcategories_highest = df['subcategory'].value_counts()
sorted_subcategory = subcategories_highest.sort_values(ascending=False)
top_subcategory = sorted_subcategory.head(1)
print ("Top subcategory with most entries:")
print(top_subcategory)

Top subcategory with most entries:
subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [24]:
# Which five clients had the most entries in the data?
clients_highest_entries = df['client_id'].value_counts()
sorted_clients = clients_highest_entries.sort_values(ascending=False)
top_5_entries= sorted_clients.head(5)
print ("Top 5 most client entries")
print(top_5_entries)

Top 5 most client entries
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [8]:
Top_5_client_ids = top_5_entries.index.tolist()
print("Client IDs of the top 5 clients", Top_5_client_ids)

Client IDs of the top 5 clients [33615, 66037, 46820, 38378, 24741]


In [25]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id =df['client_id'].value_counts().idxmax()
top_client_orders = df[df['client_id'] == top_client_id]
total_units_top_client = top_client_orders['qty'].sum()
print("Total units of the client with the most entries")
print(total_units_top_client)




Total units of the client with the most entries
64313


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  line_subtotal  \
0          17        2023  ...        7.50  105           1      115164.00   
1          20        2023  ...        1.49   21           0         523.95   
2           4        2023  ...        1.68   39           6         527.28   
3          17        2023  ...        1.23   29           3        1056.18   
4          21        2023  ...       46.43   20           1        3902.00   

   shipping_price  total_price  total_price_with_tax  line_cost  \
0         5512.50    120676.50             131839.08   80084.55   
1          312.90       836.85                914.26     316.89   
2          458.64       985.92               1077.12     306.54   
3          356.70      1412.88               1543.57     720.65   
4         6500.20     10402.20              11364.40    2163.40   

   line_cost_with_shipping  line_profit  
0                 85597.05     40591.95  
1                   629.79       519.96  
2                   765.18       679.38  
3                  1077.35       692.23  
4                  8663.60      8238.80  

[5 rows x 25 columns]

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [28]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
selected_columns = ['unit_price', 'qty', 'line_subtotal']
df_selected=df[selected_columns]
print(df_selected)
(df.head(2))


       unit_price   qty  line_subtotal
0         1096.80   105      115164.00
1           24.95    21         523.95
2           13.52    39         527.28
3           36.42    29        1056.18
4          195.10    20        3902.00
...           ...   ...            ...
54634       83.13    33        2743.29
54635      206.59    47        9709.73
54636       65.66   475       31188.50
54637        1.48   112         165.76
54638        3.01  1031        3103.31

[54639 rows x 3 columns]


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   

   order_week  order_year  ... unit_weight  qty line_number  line_subtotal  \
0          17        2023  ...        7.50  105           1      115164.00   
1          20        2023  ...        1.49   21           0         523.95   

   shipping_price  total_price  total_price_with_tax  line_cost  \
0         5512.50    120676.50             131839.08   80084.55   
1          312.90       836.85                914.26     316.89   

   line_cost_with_shipping  line_profit  
0                 85597.05     40591.95  
1                   629.79       519.96  

[2 rows x 25 columns]

In [15]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

def calculate_shipping_price(weight, qty):
    if weight * qty > 50:
        return 7 * weight * qty
    else:
        return 10 * weight * qty
    
df['shipping_price'] = df.apply(lambda row: calculate_shipping_price(row['unit_weight'], row['qty']), axis=1)
print(df)


           first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0       

In [29]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

df['total_price']= df['line_subtotal'] +df['shipping_price']
df['total_price_with_tax']= df['total_price'] * 1.0925
(df.head())

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  line_subtotal  \
0          17        2023  ...        7.50  105           1      115164.00   
1          20        2023  ...        1.49   21           0         523.95   
2           4        2023  ...        1.68   39           6         527.28   
3          17        2023  ...        1.23   29           3        1056.18   
4          21        2023  ...       46.43   20           1        3902.00   

   shipping_price  total_price  total_price_with_tax  line_cost  \
0         5512.50    120676.50             131839.08   80084.55   
1          312.90       836.85                914.26     316.89   
2          458.64       985.92               1077.12     306.54   
3          356.70      1412.88               1543.57     720.65   
4         6500.20     10402.20              11364.40    2163.40   

   line_cost_with_shipping  line_profit  
0                 85597.05     40591.95  
1                   629.79       519.96  
2                   765.18       679.38  
3                  1077.35       692.23  
4                  8663.60      8238.80  

[5 rows x 25 columns]

In [30]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost']= df['unit_cost'] * df['qty']
df['line_cost_with_shipping'] = df['line_cost'] + df['shipping_price']
(df.head())

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  line_subtotal  \
0          17        2023  ...        7.50  105           1      115164.00   
1          20        2023  ...        1.49   21           0         523.95   
2           4        2023  ...        1.68   39           6         527.28   
3          17        2023  ...        1.23   29           3        1056.18   
4          21        2023  ...       46.43   20           1        3902.00   

   shipping_price  total_price  total_price_with_tax  line_cost  \
0         5512.50    120676.50             131839.08   80084.55   
1          312.90       836.85                914.26     316.89   
2          458.64       985.92               1077.12     306.54   
3          356.70      1412.88               1543.57     720.65   
4         6500.20     10402.20              11364.40    2163.40   

   line_cost_with_shipping  line_profit  
0                 85597.05     40591.95  
1                   629.79       519.96  
2                   765.18       679.38  
3                  1077.35       692.23  
4                  8663.60      8238.80  

[5 rows x 25 columns]

In [31]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
(df.head())


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  line_subtotal  \
0          17        2023  ...        7.50  105           1      115164.00   
1          20        2023  ...        1.49   21           0         523.95   
2           4        2023  ...        1.68   39           6         527.28   
3          17        2023  ...        1.23   29           3        1056.18   
4          21        2023  ...       46.43   20           1        3902.00   

   shipping_price  total_price  total_price_with_tax  line_cost  \
0         5512.50    120676.50             131839.08   80084.55   
1          312.90       836.85                914.26     316.89   
2          458.64       985.92               1077.12     306.54   
3          356.70      1412.88               1543.57     720.65   
4         6500.20     10402.20              11364.40    2163.40   

   line_cost_with_shipping  line_profit  
0                 85597.05     40591.95  
1                   629.79       519.96  
2                   765.18       679.38  
3                  1077.35       692.23  
4                  8663.60      8238.80  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [33]:
# Check your work using the totals above
order1_total = df[df['order_id'] == 2742071]['total_price_with_tax'].sum()
order2_total = df[df['order_id'] == 2173913]['total_price_with_tax'].sum()
order3_total = df[df['order_id'] == 6128929]['total_price_with_tax'].sum()

print("Order ID 2742071 Total Price Match:", order1_total == 152811.89)
print("Order ID 2173913 Total Price Match:", order2_total == 162388.71)
print("Order ID 6128929 Total Price Match:", order3_total == 923441.25)

print("Order 2742071 total:", f"${order1_total:.2f}", "Match:", order1_total == 152811.89)
print("Order 2173913 total:", f"${order2_total:.2f}", "Match:", order2_total == 162388.71)
print("Order 6128929 total:", f"${order3_total:.2f}", "Match:", order3_total == 923441.25)

Order ID 2742071 Total Price Match: False
Order ID 2173913 Total Price Match: False
Order ID 6128929 Total Price Match: False
Order 2742071 total: $152811.90 Match: False
Order 2173913 total: $162388.72 Match: False
Order 6128929 total: $923441.24 Match: False


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [18]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995